riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1822 @@
+    std::utils::force_bool(wrap_bit_2);
+
+    // Sign extends the value in register X and stores it in register Y.
+    // Input is a 32 bit unsigned number. We check bit 7 and set all higher bits to that value.
+    instr sign_extend_byte XL, YL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> regs.mstore(0, YL, STEP + 3, tmp3_h, tmp3_l)
+        link => byte.check(X_b1)
+        link => bit7.check(tmp2_l)
+    {
+        // wrap_bit is used as sign_bit here.
+        tmp1_l = tmp2_l + wrap_bit * 2**7 + X_b1 * 2**8,
+        tmp3_h = wrap_bit * 0xffff,
+        tmp3_l = tmp2_l + wrap_bit * 0xff80
+    }
+
+    // Sign extends the value in register X and stores it in register Y.
+    // Input is a 32 bit unsigned number. We check bit 15 and set all higher bits to that value.
+    instr sign_extend_16_bits XL, YL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> regs.mstore(0, YL, STEP + 3, tmp3_h, tmp3_l)
+        link => byte.check(X_b1)
+        link => bit7.check(tmp2_l)
+    {
+        // wrap_bit is used as sign_bit here.
+        tmp1_l = X_b1 + tmp2_l * 2**8 + wrap_bit * 2**15,
+        tmp3_h = wrap_bit * 0xffff,
+        tmp3_l = tmp1_l
+    }
+
+    // ======================= assertions =========================
+
+    instr fail 
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, 0, STEP)
+    {
+      tmp1_h = 1
+    }
+
+    // Computes Q = val(Y) / val(X) and R = val(Y) % val(X) and stores them in registers Z and W.
+    instr divremu YL, XL, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, YL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, XL, STEP + 1)
+        link if (1 - XXIsZero) ~> (tmp4_h, tmp4_l, tmp3_h, tmp3_l) = arith_mul.div(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link if (1 - XXIsZero) ~> wrap_bit = add_sub.gt(tmp3_h, tmp3_l, tmp2_h, tmp2_l)
+        link if (1 - XXIsZero) ~> wrap_bit_2 = add_sub.gt(tmp3_h, tmp3_l, 0, 0)
+        link ~> regs.mstore(0, ZL, STEP + 2, tmp6_h, tmp6_l)
+        link ~> regs.mstore(0, WL, STEP + 3, tmp5_h, tmp5_l)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = tmp2_h + tmp2_l,
+
+        // if tmp2 (the divisor) is zero, remainder is set to dividend, as per RISC-V specification:
+        XXIsZero * tmp1_h + (1 - XXIsZero) * tmp3_h = tmp5_h,
+        XXIsZero * tmp1_l + (1 - XXIsZero) * tmp3_l = tmp5_l,
+
+        // remainder >= 0:
+        (1 - XXIsZero) * wrap_bit_2 = 0, 
+
+        // remainder < divisor, conditioned to val(X) not being 0:
+        (1 - XXIsZero) * (wrap_bit - 1) = 0,
+
+        // in case X is zero, we set quotient according to RISC-V specification
+        XXIsZero * 0xffff + (1 - XXIsZero) * tmp4_h = tmp6_h,
+        XXIsZero * 0xffff + (1 - XXIsZero) * tmp4_l = tmp6_l
+    }
+"# + mul_instruction
+}
+
+fn mul_instruction() -> &'static str {
+    r#"
+    // Computes V = val(X) * val(Y) and
+    // stores the lower 32 bits in register Z and the upper 32 bits in register W.
```

Comment:
```suggestion
    // stores the lower 32 bits in register ZL and the upper 32 bits in register WL.
```

Reply:
fixed all of these to use the same register name as the argument